File: disposia-engine/src/main/java/io/disposia/engine/domain/Feed.java
package io.disposia.engine.domain;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.disposia.engine.domain.FeedStatus;
import io.disposia.engine.domain.ImmutableFeed;
import org.immutables.value.Value;

import javax.annotation.Nullable;
import java.time.LocalDateTime;

@Value.Immutable
@Value.Modifiable                   // generates implementation with setters, required by mappers
@Value.Style(
    jdkOnly = true,              // prevent usage of Guava collections
    get     = {"is*", "get*"},   // Detect 'get' and 'is' prefixes in accessor methods
    init    = "set*",
    create  = "new",             // generates public no args constructor
    build   = "create"           // rename 'build' method on builder to 'create'
)
@JsonSerialize(as = ImmutableFeed.class)
@JsonDeserialize(as = ImmutableFeed.class)
public interface Feed {

    @Nullable
    String getId();

    @Nullable
    String getPodcastId();

    /*
    @Nullable
    String getExo();

    @Nullable
    String getPodcastExo();
    */

    @Nullable
    String getUrl();

    @Nullable
    LocalDateTime getLastChecked();

    @Nullable
    FeedStatus getLastStatus();

    @Nullable
    LocalDateTime getRegistrationTimestamp();

}

File: disposia-engine/src/main/scala/io/disposia/engine/domain/Feed.scala
package io.disposia.engine.domain

import java.time.LocalDateTime

case class Feed(
  id: Option[String]                           = None,
  podcastId: Option[String]                    = None,
  url: Option[String]                          = None,
  lastChecked: Option[LocalDateTime]           = None,
  lastStatus: Option[FeedStatus]               = None,
  registrationTimestamp: Option[LocalDateTime] = None,
) extends Patchable[Feed] {

  override def patch(diff: Feed): Feed = Option(diff) match {
    case None => this
    case Some(x) =>
      Feed(
        id                    = reduce(this.id, x.id),
        podcastId             = reduce(this.podcastId, x.podcastId),
        url                   = reduce(this.url, x.url),
        lastChecked           = reduce(this.lastChecked, x.lastChecked),
        lastStatus            = reduce(this.lastStatus, x.lastStatus),
        registrationTimestamp = reduce(this.registrationTimestamp, x.registrationTimestamp),
      )
  }

}
